types/src/proof/proptest_proof.rs

```diff
@@ -28,69 +27,88 @@ fn arb_accumulator_sibling() -> impl Strategy<Value = HashValue> {
+    prop_oneof![
+        arb_non_placeholder_sparse_merkle_sibling(),
+        Just(*SPARSE_MERKLE_PLACEHOLDER_HASH),
+    ]
 }
 
 impl<H> Arbitrary for AccumulatorProof<H>
 where
-    H: CryptoHasher,
+    H: CryptoHasher + 'static,
 {
     type Parameters = ();
     type Strategy = BoxedStrategy<Self>;
 
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        (0..=MAX_ACCUMULATOR_PROOF_DEPTH)
+            .prop_flat_map(|len| {
+                if len == 0 {
+                    Just(vec![]).boxed()
+                } else {
+                    (
+                        arb_non_placeholder_accumulator_sibling(),
+                        vec(arb_accumulator_sibling(), 0..len),
+                    )
+                        .prop_map(|(first_sibling, other_siblings)| {
+                            let mut siblings = vec![first_sibling];
+                            siblings.extend(other_siblings.into_iter());
+                            siblings
+                        })
+                        .boxed()
+                }
+            })
+            .prop_map(AccumulatorProof::<H>::new)
+            .boxed()
+    }
+}
+
+impl Arbitrary for SparseMerkleProof {
+    type Parameters = ();
+    type Strategy = BoxedStrategy<Self>;
+
     fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
         (
-            arb_non_placeholder_accumulator_sibling(),
-            vec(arb_accumulator_sibling(), MAX_ACCUMULATOR_PROOF_DEPTH - 1),
+            any::<Option<(HashValue, HashValue)>>(),
+            (0..=256usize).prop_flat_map(|len| {
+                if len == 0 {
+                    Just(vec![]).boxed()
+                } else {
+                    (
+                        vec(arb_sparse_merkle_sibling(), 0..len),
+                        arb_non_placeholder_sparse_merkle_sibling(),
+                    )
+                        .prop_map(|(other_siblings, last_sibling)| {
+                            let mut siblings = other_siblings;
+                            siblings.push(last_sibling);
+                            siblings
+                        })
+                        .boxed()
+                }
+            }),
         )
-            .prop_map(|(first_sibling, other_siblings)| {
-                let mut siblings = vec![first_sibling];
-                siblings.extend(other_siblings.into_iter());
-                AccumulatorProof::<H>::new(siblings)
-            })
+            .prop_map(|(leaf, siblings)| SparseMerkleProof::new(leaf, siblings))
             .boxed()
     }
 }
 
-impl_arbitrary_for_proof!(SparseMerkleProof, arb_sparse_merkle_proof);
-impl_arbitrary_for_proof!(
-    AccumulatorConsistencyProof,
-    arb_accumulator_consistency_proof
-);
+impl Arbitrary for AccumulatorConsistencyProof {
+    type Parameters = ();
+    type Strategy = BoxedStrategy<Self>;
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
```

Comment:
Why don't we cover the case of empty proof here?

Reply:
This does include empty proof? See `0..` below.

Reply:
I must be blind.